src/Plugin/Action/GenerateImageDerivative.php

```diff
@@ -0,0 +1,218 @@
+   * @param \Drupal\islandora\IslandoraUtils $utils
+   *   Islandora utility functions.
+   */
+  public function __construct(
+    array $configuration,
+    $plugin_id,
+    $plugin_definition,
+    AccountInterface $account,
+    EntityTypeManager $entity_type_manager,
+    EventGeneratorInterface $event_generator,
+    StatefulStomp $stomp,
+    JwtAuth $auth,
+    IslandoraUtils $utils,
+    MediaSourceService $media_source
+  ) {
+    parent::__construct(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $account,
+      $entity_type_manager,
+      $event_generator,
+      $stomp,
+      $auth
+    );
+    $this->utils = $utils;
+    $this->mediaSource = $media_source;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('current_user'),
+      $container->get('entity_type.manager'),
+      $container->get('islandora.eventgenerator'),
+      $container->get('islandora.stomp'),
+      $container->get('jwt.authentication.jwt'),
+      $container->get('islandora.utils'),
+      $container->get('islandora.media_source_service')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function defaultConfiguration() {
+    return [
+      'queue' => 'islandora-connector-houdini',
+      'event' => 'Generate Derivative',
+      'source_term_uri' => '',
+      'derivative_term_uri' => '',
+      'mimetype' => 'image/jpeg',
+      'args' => '',
+    ];
+  }
+
+  protected function generateData($entity) {
+    $data = parent::generateData($entity);
+    
+    // Find media belonging to node that has the source term, and get its file url.
+    $source_term = $this->utils->getTermForUri($this->configuration['source_term_uri']);
+    $source_media = $this->utils->getMediaWithTerm($entity, $source_term);
+    $source_field = $this->mediaSource->getSourceFieldName($source_media->bundle());
+    $files = $source_media->get($source_field)->referencedEntities();
+    $file = reset($files);
+    $data['source_uri'] = $file->url('canonical', ['absolute' => TRUE]);
+
+    $derivative_term = $this->utils->getTermForUri($this->configuration['derivative_term_uri']);
+    $route_params = ['node' => $entity->id(), 'media_type' => 'image', 'taxonomy_term' => $derivative_term->id()];
+    $data['destination_uri'] = Url::fromRoute('islandora.media_source_put_to_node', $route_params)
+      ->setAbsolute()
+      ->toString();
+
+    $parts = explode('/', $data['mimetype']);
+    if (count($parts) > 1) {
+      $extension = $parts[1];
+    } elseif (!empty($data['mimetype'])) {
+      $extension = $data['mimetype'];
+    } else {
+      $extension = "jpeg";
+    }
+
+    $parts = explode('#', $this->configuration['derivative_term_uri']);
+    if (count($parts) > 1) {
+      $name = $entity->id() . ' - ' . $parts[1];
```

Comment:
$entity->uuid() to avoid collisions? Should it be the uuid of the media entity?(again, maybe not understanding this)

Reply:
Good call. We might run into that on multisites.